src/systems/system.C

```diff
@@ -342,18 +344,28 @@ void System::init_matrices ()
       // want to attach the same matrix to the DofMap twice
       if (!this->get_dof_map().is_attached(m))
         this->get_dof_map().attach_matrix(m);
+
+      if (!pr.second->assembly_preference_set())
+        {
+          const bool use_hash =
+              this->_prefer_hash_table_matrix_assembly && pr.second->supports_hash_table();
+          pr.second->use_hash_table(use_hash);
+          if (!use_hash)
+            this->_require_sparsity_pattern = true;
+        }
```

Comment:
It looks like we definitely need some test coverage of a system with a mix of hash and non-hash matrices.  Even if it's not hit in CI until we upgrade PETSc there, we'll at least get some early warning on our own machines.

Reply:
We already have a PETSc sufficient for testing in apptainer. I will look into making something

Reply:
I still need to add some testing for this